src/super_gradients/training/sg_model/sg_model.py

```diff
@@ -320,6 +313,87 @@ def build_model(self,  # noqa: C901 - too complex
+            (self.experiment_name if none is given) to resume KD training
+        :param kd_model_external_checkpoint_path: The path to the external checkpoint to be loaded. Can be absolute or relative
+                                           (ie: path/to/checkpoint.pth). If provided, will automatically attempt to
+                                           load the checkpoint even if the load_checkpoint flag is not provided
+        :param freeze_teacher_eval_mode:   Whether to freeze self.teacher at eval mode regardless of self.train(mode)
+        :return:
+        """
+
+        kd_model_arch_params = {}
+
+        if 'num_classes' not in student_arch_params.keys():
+            if self.dataset_interface is None:
+                raise Exception('Error', 'Number of classes not defined in arch params and dataset is not defined')
+            else:
+                student_arch_params['num_classes'] = len(self.classes)
+
+        kd_model_arch_params['num_classes'] = student_arch_params['num_classes']
+        teacher_arch_params['num_classes'] = student_arch_params['num_classes']
+
+        student_arch_params = core_utils.HpmStruct(**student_arch_params)
+        teacher_arch_params = core_utils.HpmStruct(**teacher_arch_params)
+
+        student_net, _ = sg_model_utils.instantiate_net(student_architecture, student_arch_params)
+        teacher_net, _ = sg_model_utils.instantiate_net(teacher_architecture, teacher_arch_params)
+
+        if teacher_checkpoint_path is not None:
+            load_teachers_ema = 'ema_net' in read_ckpt_state_dict(teacher_checkpoint_path).keys()
+            load_checkpoint_to_model(ckpt_local_path=teacher_checkpoint_path,
+                                     load_backbone=False,
+                                     net=teacher_net,
+                                     strict='no_key_matching',
+                                     load_weights_only=True,
+                                     load_ema_as_net=load_teachers_ema)
+
+        teacher_pretrained_weights = core_utils.get_param(teacher_arch_params, 'pretrained_weights', default_val=None)
+
+        if teacher_pretrained_weights is not None:
+            pretrained_num_classes = PRETRAINED_NUM_CLASSES[teacher_pretrained_weights]
+            if pretrained_num_classes != kd_model_arch_params['num_classes']:
+                raise ValueError("Pretrained dataset number of classes must be equal to the teacher and student's "
+                                 "number of classes.")
+            if teacher_checkpoint_path is not None:
+                if teacher_pretrained_weights:
+                    logger.warning(
+                        teacher_checkpoint_path + " checkpoint is "
+                                                  "overriding " + teacher_pretrained_weights + "for teacher model")
+
+            load_pretrained_weights(teacher_net, teacher_architecture, teacher_pretrained_weights)
+
+        architecture = KDModule(student=student_net,
```

Comment:
At this point if neither pretrained_weights nor teacher_checkpoint_path were provided the teacher will start from scratch. This should be asserted IMO

Reply:
Good point, will fix it.

Reply:
fixed